runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -156,6 +154,172 @@ static TR::Node *lowerCASValues(
+
+   // insert store of constant 1
+   // the value must go wherever the value returned by the helper call goes
+   // so that the code in the target block picks up the constant if we fast-path
+   // (i.e. jump around) the call
+   auto* anchoredNode = anchoredCallTT->getNode()->getFirstChild(); // call node is under a treetop node
+   traceMsg(comp, "<LEO> Anchored call has been transformed into %s node n%dn\n",
+            anchoredNode->getOpCode().getName(), anchoredNode->getGlobalIndex());
+   auto* const1Node = TR::Node::iconst(1);
+   TR::Node* storeNode = NULL;
+   if (anchoredNode->getOpCodeValue() == TR::iRegLoad)
+      {
+      traceMsg(comp, "<LEO> Storing constant 1 in register %s\n",
+               comp->getDebug()->getGlobalRegisterName(anchoredNode->getGlobalRegisterNumber()));
+      storeNode = TR::Node::create(TR::iRegStore, 1, const1Node);
+      storeNode->setGlobalRegisterNumber(anchoredNode->getGlobalRegisterNumber());
+      }
+   else if (anchoredNode->getOpCodeValue() == TR::iload)
+      {
+      traceMsg(comp, "<LEO> Storing constant 1 to symref %d (%s)\n",
+               anchoredNode->getSymbolReference()->getReferenceNumber(), anchoredNode->getSymbolReference()->getName(comp->getDebug()));
+      storeNode = TR::Node::create(TR::istore, 1, const1Node);
+      storeNode->setSymbolReference(anchoredNode->getSymbolReference());
+      }
+   else
+      TR_ASSERT_FATAL(false, "Anchord call has been turned into unexpected opcode %s\n", anchoredNode->getOpCode().getName());
+   prevBlock->append(TR::TreeTop::create(comp, storeNode));
+   comp->dumpMethodTrees("Trees after inserting store of constant 1");
+
+   // instert acmpeq for fastpath, taking care to set the proper register dependencies
+   auto* ifacmpeqNode = TR::Node::createif(TR::ifacmpeq, anchoredCallArg1TT->getNode()->getFirstChild(), anchoredCallArg2TT->getNode()->getFirstChild(), targetBlock->getEntry());
+   if (anchoredNode->getOpCodeValue() == TR::iRegLoad)
+      {
+      auto* depNode = TR::Node::create(TR::PassThrough, 1, storeNode->getChild(0));
+      depNode->setGlobalRegisterNumber(storeNode->getGlobalRegisterNumber());
+
+      TR::Node* glRegDeps = TR::Node::create(TR::GlRegDeps);
+      glRegDeps->addChildren(&depNode, 1);
+      ifacmpeqNode->addChildren(&glRegDeps, 1);
+
+      if (callBlock->getExit()->getNode()->getNumChildren() > 0)
+         {
+         auto* expectedDeps = callBlock->getExit()->getNode()->getFirstChild();
+         for (int i = 0; i < expectedDeps->getNumChildren(); ++i)
+            {
+            TR::Node* temp = expectedDeps->getChild(i);
+            if (temp->getGlobalRegisterNumber() == depNode->getGlobalRegisterNumber())
+               continue;
+            glRegDeps->addChildren(&temp, 1);
+            }
+         }
+      }
+   prevBlock->append(TR::TreeTop::create(comp, ifacmpeqNode));
+
+   comp->dumpMethodTrees("Trees after inserting acmpeq fast path");
+   }
```

Comment:
I'd appreciate reviews of this code from @r30shah @andrewcraik @hzongaro and @aviansie-ben. I'm not really happy with this code, but I haven't settled on a good way to clean it up yet. In particular, I don't like the way GlRegDeps are handled right now. I'm open any suggestions people might have.

Reply:
I've discovered and resolved a bug in this code so the above listing is now out of date.